middleware/file/include/file/message.h

```diff
@@ -0,0 +1,111 @@
+//!
+//! Copyright (c) 2024, The casual project
+//!
+//! This software is licensed under the MIT license, https://opensource.org/licenses/MIT
+//!
+
+#pragma once
+
+#include "common/message/type.h"
+#include "common/serialize/macro.h"
+#include "common/code/file.h"
+
+#include "common/process.h"
+#include "common/transaction/id.h"
+#include "common/transaction/context.h"
+
+#include <filesystem>
+
+namespace casual
+{
+   namespace file::message
+   {
+      namespace reserve
+      {
+         using base_request = common::message::basic_request< common::message::Type::file_reserve_request>;
+         struct Request : base_request
+         {
+            using base_request::base_request;
+
+            common::transaction::ID trid;
+            std::filesystem::path path;
+            bool wait{true};
+
+            CASUAL_CONST_CORRECT_SERIALIZE(
+               base_request::serialize( archive);
+               CASUAL_SERIALIZE( trid);
+               CASUAL_SERIALIZE( path);
+               CASUAL_SERIALIZE( wait);
+            )
+         };
+
+         using base_reply = common::message::basic_reply< common::message::Type::file_reserve_reply>;
+         struct Reply : base_reply
+         {
+            std::filesystem::path path;
+            common::code::file code{};
+
+            CASUAL_CONST_CORRECT_SERIALIZE(
+               base_reply::serialize( archive);
+               CASUAL_SERIALIZE( path);
+               CASUAL_SERIALIZE( code);
+            )
+         };
+
+
+         namespace create
+         {
+            inline namespace blocking
+            {
+               inline auto request( std::filesystem::path path)
+               {
+                  auto& transaction = common::transaction::context().current();
+
+                  if( transaction)
+                  {
+                     // there's a transaction so make sure to trigger an interaction with the TM
+                     transaction.external();
+                  }
+                  else
+                  {
+                     common::code::raise::error( common::code::casual::preconditions, "no transaction");
+                  }
```

Comment:
I don't think _message_ should be responsible for the validity of it's content. I think you should move this to some other place. And get rid of the include of transaction context and such.